{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from analytics.network_generation import generate_comention_network\n",
    "from analytics.data_preparation import (mentions_to_occurrence,\n",
    "                                        is_experiment_related,\n",
    "                                        dummy_clean_up,\n",
    "                                        is_not_single_letter)\n",
    "from analytics.metrics import (compute_degree_centrality,\n",
    "                               compute_pagerank_centrality,\n",
    "                               compute_betweenness_centrality,\n",
    "                               detect_communities,\n",
    "                               compute_all_metrics)\n",
    "from analytics.export import (save_nodes,\n",
    "                              save_to_gephi)\n",
    "from analytics.paths import (minimum_spanning_tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Read mentions data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load 10000 lines of the mention data sample\n",
    "mentions = pd.read_csv(\"data/mention_data_sample.csv\", nrows=10000)\n",
    "\n",
    "# Extract unique paper/seciton/paragraph identifiers\n",
    "mentions[\"paper\"] = mentions[\"paper_id\"].apply(\n",
    "    lambda x: x.split(\":\")[0])\n",
    "mentions[\"section\"] = mentions[\"paper_id\"].apply(\n",
    "    lambda x: \":\".join([x.split(\":\")[0], x.split(\":\")[1]]))\n",
    "\n",
    "mentions = mentions.rename(columns={\"paper_id\": \"paragraph\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entity</th>\n",
       "      <th>paragraph</th>\n",
       "      <th>paper</th>\n",
       "      <th>section</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tract</td>\n",
       "      <td>1:Introduction:2</td>\n",
       "      <td>1</td>\n",
       "      <td>1:Introduction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nasopharyngeal</td>\n",
       "      <td>1:Data Collection ::: Methods:4</td>\n",
       "      <td>1</td>\n",
       "      <td>1:Data Collection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tract</td>\n",
       "      <td>1:Data Collection ::: Methods:4</td>\n",
       "      <td>1</td>\n",
       "      <td>1:Data Collection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>lung</td>\n",
       "      <td>1:Results:10</td>\n",
       "      <td>1</td>\n",
       "      <td>1:Results</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>heart</td>\n",
       "      <td>1:Results:10</td>\n",
       "      <td>1</td>\n",
       "      <td>1:Results</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>A549 cells</td>\n",
       "      <td>19:Caption:44</td>\n",
       "      <td>19</td>\n",
       "      <td>19:Caption</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>hNTH</td>\n",
       "      <td>19:Caption:44</td>\n",
       "      <td>19</td>\n",
       "      <td>19:Caption</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>A549 cells</td>\n",
       "      <td>19:Caption:44</td>\n",
       "      <td>19</td>\n",
       "      <td>19:Caption</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9998</th>\n",
       "      <td>hNTH</td>\n",
       "      <td>19:Caption:44</td>\n",
       "      <td>19</td>\n",
       "      <td>19:Caption</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9999</th>\n",
       "      <td>vector control cells</td>\n",
       "      <td>19:Caption:44</td>\n",
       "      <td>19</td>\n",
       "      <td>19:Caption</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10000 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    entity                        paragraph paper  \\\n",
       "0                    tract                 1:Introduction:2     1   \n",
       "1           nasopharyngeal  1:Data Collection ::: Methods:4     1   \n",
       "2                    tract  1:Data Collection ::: Methods:4     1   \n",
       "3                     lung                     1:Results:10     1   \n",
       "4                    heart                     1:Results:10     1   \n",
       "...                    ...                              ...   ...   \n",
       "9995            A549 cells                    19:Caption:44    19   \n",
       "9996                  hNTH                    19:Caption:44    19   \n",
       "9997            A549 cells                    19:Caption:44    19   \n",
       "9998                  hNTH                    19:Caption:44    19   \n",
       "9999  vector control cells                    19:Caption:44    19   \n",
       "\n",
       "                 section  \n",
       "0         1:Introduction  \n",
       "1     1:Data Collection   \n",
       "2     1:Data Collection   \n",
       "3              1:Results  \n",
       "4              1:Results  \n",
       "...                  ...  \n",
       "9995          19:Caption  \n",
       "9996          19:Caption  \n",
       "9997          19:Caption  \n",
       "9998          19:Caption  \n",
       "9999          19:Caption  \n",
       "\n",
       "[10000 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mentions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Transform mentions into occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cleaning up the entities...\n",
      "Aggregating occurrences of entities....\n",
      "Saving the occurrence data....\n"
     ]
    }
   ],
   "source": [
    "occurrence_data, counts = mentions_to_occurrence(\n",
    "    mentions,\n",
    "    term_column=\"entity\",\n",
    "    factor_columns=[\"paper\", \"section\", \"paragraph\"],\n",
    "    term_cleanup=dummy_clean_up,\n",
    "    term_filter=is_not_single_letter,\n",
    "    mention_filter=lambda data: ~data[\"section\"].apply(is_experiment_related),\n",
    "    dump_prefix=\"data/example_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paragraph</th>\n",
       "      <th>paper</th>\n",
       "      <th>section</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1 nucleotide</th>\n",
       "      <td>{6:The Discontinuous Step In Nidovirus Sg Rna ...</td>\n",
       "      <td>{6}</td>\n",
       "      <td>{6:The Discontinuous Step In Nidovirus Sg Rna ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1,2,9</th>\n",
       "      <td>{19:Background:4}</td>\n",
       "      <td>{19}</td>\n",
       "      <td>{19:Background}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10/100,000</th>\n",
       "      <td>{8:Acute Respiratory Distress Syndrome:14}</td>\n",
       "      <td>{8}</td>\n",
       "      <td>{8:Acute Respiratory Distress Syndrome}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>{9:Pulmonary Vascular Disease ::: Review:23}</td>\n",
       "      <td>{9}</td>\n",
       "      <td>{9:Pulmonary Vascular Disease }</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11c</th>\n",
       "      <td>{14:A10L And L4R Associate With Microtubules I...</td>\n",
       "      <td>{14}</td>\n",
       "      <td>{14:A10L And L4R Associate With Microtubules I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>–208</th>\n",
       "      <td>{5:Syncytia Formation:17}</td>\n",
       "      <td>{5}</td>\n",
       "      <td>{5:Syncytia Formation}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>–met</th>\n",
       "      <td>{11:In Vitro Transcription And Translation :::...</td>\n",
       "      <td>{11}</td>\n",
       "      <td>{11:In Vitro Transcription And Translation }</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>†in</th>\n",
       "      <td>{7:Caption:27}</td>\n",
       "      <td>{7}</td>\n",
       "      <td>{7:Caption}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>∼250 kda</th>\n",
       "      <td>{12:The Mechanism Of Dominant-Negative Inhibit...</td>\n",
       "      <td>{12}</td>\n",
       "      <td>{12:Discussion, 12:The Mechanism Of Dominant-N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>∼80</th>\n",
       "      <td>{11:Identification Of An Antizyme Gene In Schi...</td>\n",
       "      <td>{12, 11}</td>\n",
       "      <td>{11:Identification Of An Antizyme Gene In Schi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2880 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      paragraph     paper  \\\n",
       "entity                                                                      \n",
       "1 nucleotide  {6:The Discontinuous Step In Nidovirus Sg Rna ...       {6}   \n",
       "1,2,9                                         {19:Background:4}      {19}   \n",
       "10/100,000           {8:Acute Respiratory Distress Syndrome:14}       {8}   \n",
       "104                {9:Pulmonary Vascular Disease ::: Review:23}       {9}   \n",
       "11c           {14:A10L And L4R Associate With Microtubules I...      {14}   \n",
       "...                                                         ...       ...   \n",
       "–208                                  {5:Syncytia Formation:17}       {5}   \n",
       "–met          {11:In Vitro Transcription And Translation :::...      {11}   \n",
       "†in                                              {7:Caption:27}       {7}   \n",
       "∼250 kda      {12:The Mechanism Of Dominant-Negative Inhibit...      {12}   \n",
       "∼80           {11:Identification Of An Antizyme Gene In Schi...  {12, 11}   \n",
       "\n",
       "                                                        section  \n",
       "entity                                                           \n",
       "1 nucleotide  {6:The Discontinuous Step In Nidovirus Sg Rna ...  \n",
       "1,2,9                                           {19:Background}  \n",
       "10/100,000              {8:Acute Respiratory Distress Syndrome}  \n",
       "104                             {9:Pulmonary Vascular Disease }  \n",
       "11c           {14:A10L And L4R Associate With Microtubules I...  \n",
       "...                                                         ...  \n",
       "–208                                     {5:Syncytia Formation}  \n",
       "–met               {11:In Vitro Transcription And Translation }  \n",
       "†in                                                 {7:Caption}  \n",
       "∼250 kda      {12:Discussion, 12:The Mechanism Of Dominant-N...  \n",
       "∼80           {11:Identification Of An Antizyme Gene In Schi...  \n",
       "\n",
       "[2880 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occurrence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paragraph</th>\n",
       "      <th>paper</th>\n",
       "      <th>section</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>entity</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16,20</th>\n",
       "      <td>{2:Protein Tyrosine Nitration In The Lung: Doe...</td>\n",
       "      <td>{2}</td>\n",
       "      <td>{2:Injurious Properties Of No•, 2:Protein Tyro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2-oxoglutarate</th>\n",
       "      <td>{20:Background:2, 20:The Viral Alkb Domains Ar...</td>\n",
       "      <td>{20}</td>\n",
       "      <td>{20:The Viral Alkb Domains Are Most Likely Fun...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2c cells</th>\n",
       "      <td>{25:Rna Isolation And Differential Display Pcr...</td>\n",
       "      <td>{25}</td>\n",
       "      <td>{25:Discussion, 25:Rna Isolation And Different...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-methylcytosine</th>\n",
       "      <td>{20:Background:2, 20:The Alkb Domain Probably ...</td>\n",
       "      <td>{20}</td>\n",
       "      <td>{20:The Alkb Domain Probably Protects Virus Rn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3-nitrotyrosine</th>\n",
       "      <td>{2:Protein Tyrosine Nitration In The Lung: Doe...</td>\n",
       "      <td>{2}</td>\n",
       "      <td>{2:Injurious Properties Of No•, 2:Protein Tyro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>β-galactosidase</th>\n",
       "      <td>{11:Identification Of Antizyme Genes In Nemato...</td>\n",
       "      <td>{11}</td>\n",
       "      <td>{11:Identification Of Antizyme Genes In Nemato...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>γ-tubulin</th>\n",
       "      <td>{14:Vaccinia Virus Infection Disrupts Centroso...</td>\n",
       "      <td>{14}</td>\n",
       "      <td>{14:Vaccinia Virus Infection Disrupts Centroso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>–20</th>\n",
       "      <td>{12:Immunofluorescence Staining ::: Materials ...</td>\n",
       "      <td>{14, 12}</td>\n",
       "      <td>{14:Antibodies And Immunofluorescence Microsco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>∼250 kda</th>\n",
       "      <td>{12:The Mechanism Of Dominant-Negative Inhibit...</td>\n",
       "      <td>{12}</td>\n",
       "      <td>{12:Discussion, 12:The Mechanism Of Dominant-N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>∼80</th>\n",
       "      <td>{11:Identification Of An Antizyme Gene In Schi...</td>\n",
       "      <td>{12, 11}</td>\n",
       "      <td>{11:Identification Of An Antizyme Gene In Schi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          paragraph     paper  \\\n",
       "entity                                                                          \n",
       "16,20             {2:Protein Tyrosine Nitration In The Lung: Doe...       {2}   \n",
       "2-oxoglutarate    {20:Background:2, 20:The Viral Alkb Domains Ar...      {20}   \n",
       "2c cells          {25:Rna Isolation And Differential Display Pcr...      {25}   \n",
       "3-methylcytosine  {20:Background:2, 20:The Alkb Domain Probably ...      {20}   \n",
       "3-nitrotyrosine   {2:Protein Tyrosine Nitration In The Lung: Doe...       {2}   \n",
       "...                                                             ...       ...   \n",
       "β-galactosidase   {11:Identification Of Antizyme Genes In Nemato...      {11}   \n",
       "γ-tubulin         {14:Vaccinia Virus Infection Disrupts Centroso...      {14}   \n",
       "–20               {12:Immunofluorescence Staining ::: Materials ...  {14, 12}   \n",
       "∼250 kda          {12:The Mechanism Of Dominant-Negative Inhibit...      {12}   \n",
       "∼80               {11:Identification Of An Antizyme Gene In Schi...  {12, 11}   \n",
       "\n",
       "                                                            section  \n",
       "entity                                                               \n",
       "16,20             {2:Injurious Properties Of No•, 2:Protein Tyro...  \n",
       "2-oxoglutarate    {20:The Viral Alkb Domains Are Most Likely Fun...  \n",
       "2c cells          {25:Discussion, 25:Rna Isolation And Different...  \n",
       "3-methylcytosine  {20:The Alkb Domain Probably Protects Virus Rn...  \n",
       "3-nitrotyrosine   {2:Injurious Properties Of No•, 2:Protein Tyro...  \n",
       "...                                                             ...  \n",
       "β-galactosidase   {11:Identification Of Antizyme Genes In Nemato...  \n",
       "γ-tubulin         {14:Vaccinia Virus Infection Disrupts Centroso...  \n",
       "–20               {14:Antibodies And Immunofluorescence Microsco...  \n",
       "∼250 kda          {12:Discussion, 12:The Mechanism Of Dominant-N...  \n",
       "∼80               {11:Identification Of An Antizyme Gene In Schi...  \n",
       "\n",
       "[896 rows x 3 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter entities that occur only once (only in one paragraph, usually represent noisy terms)\n",
    "occurrence_data = occurrence_data[occurrence_data[\"paragraph\"].apply(lambda x: len(x) > 1)]\n",
    "occurrence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'paper': 41, 'section': 465, 'paragraph': 1045}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Generate co-occurrence networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Examining 400960 pairs of terms for co-occurrence...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed 1 (0%) pairs     \u001b[1K\r",
      "Processed 2 (0%) pairs     \u001b[1K\r",
      "Processed 3 (0%) pairs     \u001b[1K\r",
      "Processed 4 (0%) pairs     \u001b[1K\r",
      "Processed 5 (0%) pairs     \u001b[1K\r",
      "Processed 6 (0%) pairs     \u001b[1K\r",
      "Processed 7 (0%) pairs     \u001b[1K\r",
      "Processed 8 (0%) pairs     \u001b[1K\r",
      "Processed 9 (0%) pairs     \u001b[1K\r",
      "Processed 10 (0%) pairs     \u001b[1K\r",
      "Processed 11 (0%) pairs     \u001b[1K\r",
      "Processed 12 (0%) pairs     \u001b[1K\r",
      "Processed 13 (0%) pairs     \u001b[1K\r",
      "Processed 14 (0%) pairs     \u001b[1K\r",
      "Processed 15 (0%) pairs     \u001b[1K\r",
      "Processed 16 (0%) pairs     \u001b[1K\r",
      "Processed 17 (0%) pairs     \u001b[1K\r",
      "Processed 18 (0%) pairs     \u001b[1K\r",
      "Processed 19 (0%) pairs     \u001b[1K\r",
      "Processed 20 (0%) pairs     \u001b[1K\r",
      "Processed 21 (0%) pairs     \u001b[1K\r",
      "Processed 22 (0%) pairs     \u001b[1K\r",
      "Processed 23 (0%) pairs     \u001b[1K\r",
      "Processed 24 (0%) pairs     \u001b[1K\r",
      "Processed 25 (0%) pairs     \u001b[1K\r",
      "Processed 26 (0%) pairs     \u001b[1K\r",
      "Processed 27 (0%) pairs     \u001b[1K\r",
      "Processed 28 (0%) pairs     \u001b[1K\r",
      "Processed 29 (0%) pairs     \u001b[1K\r",
      "Processed 30 (0%) pairs     \u001b[1K\r",
      "Processed 31 (0%) pairs     \u001b[1K\r",
      "Processed 32 (0%) pairs     \u001b[1K\r",
      "Processed 33 (0%) pairs     \u001b[1K\r",
      "Processed 34 (0%) pairs     \u001b[1K\r",
      "Processed 35 (0%) pairs     \u001b[1K\r",
      "Processed 36 (0%) pairs     \u001b[1K\r",
      "Processed 37 (0%) pairs     \u001b[1K\r",
      "Processed 38 (0%) pairs     \u001b[1K\r",
      "Processed 39 (0%) pairs     \u001b[1K\r",
      "Processed 40 (0%) pairs     \u001b[1K\r",
      "Processed 41 (0%) pairs     \u001b[1K\r",
      "Processed 42 (0%) pairs     \u001b[1K\r",
      "Processed 43 (0%) pairs     \u001b[1K\r",
      "Processed 44 (0%) pairs     \u001b[1K\r",
      "Processed 45 (0%) pairs     \u001b[1K\r",
      "Processed 46 (0%) pairs     \u001b[1K\r",
      "Processed 47 (0%) pairs     \u001b[1K\r",
      "Processed 48 (0%) pairs     \u001b[1K\r",
      "Processed 49 (0%) pairs     \u001b[1K\r",
      "Processed 50 (0%) pairs     \u001b[1K\r",
      "Processed 51 (0%) pairs     \u001b[1K\r",
      "Processed 52 (0%) pairs     \u001b[1K\r",
      "Processed 53 (0%) pairs     \u001b[1K\r",
      "Processed 54 (0%) pairs     \u001b[1K\r",
      "Processed 55 (0%) pairs     \u001b[1K\r",
      "Processed 56 (0%) pairs     \u001b[1K\r",
      "Processed 57 (0%) pairs     \u001b[1K\r",
      "Processed 58 (0%) pairs     \u001b[1K\r",
      "Processed 59 (0%) pairs     \u001b[1K\r",
      "Processed 60 (0%) pairs     \u001b[1K\r",
      "Processed 61 (0%) pairs     \u001b[1K\r",
      "Processed 62 (0%) pairs     \u001b[1K\r",
      "Processed 63 (0%) pairs     \u001b[1K\r",
      "Processed 64 (0%) pairs     \u001b[1K\r",
      "Processed 65 (0%) pairs     \u001b[1K\r",
      "Processed 66 (0%) pairs     \u001b[1K\r",
      "Processed 67 (0%) pairs     \u001b[1K\r",
      "Processed 68 (0%) pairs     \u001b[1K\r",
      "Processed 69 (0%) pairs     \u001b[1K\r",
      "Processed 70 (0%) pairs     \u001b[1K\r",
      "Processed 71 (0%) pairs     \u001b[1K\r",
      "Processed 72 (0%) pairs     \u001b[1K\r",
      "Processed 73 (0%) pairs     \u001b[1K\r",
      "Processed 74 (0%) pairs     \u001b[1K\r",
      "Processed 75 (0%) pairs     \u001b[1K\r",
      "Processed 76 (0%) pairs     \u001b[1K\r",
      "Processed 77 (0%) pairs     \u001b[1K\r",
      "Processed 78 (0%) pairs     \u001b[1K\r",
      "Processed 79 (0%) pairs     \u001b[1K\r",
      "Processed 80 (0%) pairs     \u001b[1K\r",
      "Processed 81 (0%) pairs     \u001b[1K\r",
      "Processed 82 (0%) pairs     \u001b[1K\r",
      "Processed 83 (0%) pairs     \u001b[1K\r",
      "Processed 84 (0%) pairs     \u001b[1K\r",
      "Processed 85 (0%) pairs     \u001b[1K\r",
      "Processed 86 (0%) pairs     \u001b[1K\r",
      "Processed 87 (0%) pairs     \u001b[1K\r",
      "Processed 88 (0%) pairs     \u001b[1K\r",
      "Processed 89 (0%) pairs     \u001b[1K\r",
      "Processed 90 (0%) pairs     \u001b[1K\r",
      "Processed 91 (0%) pairs     \u001b[1K\r",
      "Processed 92 (0%) pairs     \u001b[1K\r",
      "Processed 93 (0%) pairs     \u001b[1K\r",
      "Processed 94 (0%) pairs     \u001b[1K\r",
      "Processed 95 (0%) pairs     \u001b[1K\r",
      "Processed 96 (0%) pairs     \u001b[1K\r",
      "Processed 97 (0%) pairs     \u001b[1K\r",
      "Processed 98 (0%) pairs     \u001b[1K\r",
      "Processed 99 (0%) pairs     \u001b[1K\r",
      "Processed 100 (0%) pairs     \u001b[1K\r",
      "Processed 101 (0%) pairs     \u001b[1K\r",
      "Processed 102 (0%) pairs     \u001b[1K\r",
      "Processed 103 (0%) pairs     \u001b[1K\r",
      "Processed 104 (0%) pairs     \u001b[1K\r",
      "Processed 105 (0%) pairs     \u001b[1K\r",
      "Processed 106 (0%) pairs     \u001b[1K\r",
      "Processed 107 (0%) pairs     \u001b[1K\r",
      "Processed 108 (0%) pairs     \u001b[1K\r",
      "Processed 109 (0%) pairs     \u001b[1K\r",
      "Processed 110 (0%) pairs     \u001b[1K\r",
      "Processed 111 (0%) pairs     \u001b[1K\r",
      "Processed 112 (0%) pairs     \u001b[1K\r",
      "Processed 113 (0%) pairs     \u001b[1K\r",
      "Processed 114 (0%) pairs     \u001b[1K\r",
      "Processed 115 (0%) pairs     \u001b[1K\r",
      "Processed 116 (0%) pairs     \u001b[1K\r",
      "Processed 117 (0%) pairs     \u001b[1K\r",
      "Processed 118 (0%) pairs     \u001b[1K\r",
      "Processed 119 (0%) pairs     \u001b[1K\r",
      "Processed 120 (0%) pairs     \u001b[1K\r",
      "Processed 121 (0%) pairs     \u001b[1K\r",
      "Processed 122 (0%) pairs     \u001b[1K\r",
      "Processed 123 (0%) pairs     \u001b[1K\r",
      "Processed 124 (0%) pairs     \u001b[1K\r",
      "Processed 125 (0%) pairs     \u001b[1K\r",
      "Processed 126 (0%) pairs     \u001b[1K\r",
      "Processed 127 (0%) pairs     \u001b[1K\r",
      "Processed 128 (0%) pairs     \u001b[1K\r",
      "Processed 129 (0%) pairs     \u001b[1K\r",
      "Processed 130 (0%) pairs     \u001b[1K\r",
      "Processed 131 (0%) pairs     \u001b[1K\r",
      "Processed 132 (0%) pairs     \u001b[1K\r",
      "Processed 133 (0%) pairs     \u001b[1K\r",
      "Processed 134 (0%) pairs     \u001b[1K\r",
      "Processed 135 (0%) pairs     \u001b[1K\r",
      "Processed 136 (0%) pairs     \u001b[1K\r",
      "Processed 137 (0%) pairs     \u001b[1K\r",
      "Processed 138 (0%) pairs     \u001b[1K\r",
      "Processed 139 (0%) pairs     \u001b[1K\r",
      "Processed 140 (0%) pairs     \u001b[1K\r",
      "Processed 141 (0%) pairs     \u001b[1K\r",
      "Processed 142 (0%) pairs     \u001b[1K\r",
      "Processed 143 (0%) pairs     \u001b[1K\r",
      "Processed 144 (0%) pairs     \u001b[1K\r",
      "Processed 145 (0%) pairs     \u001b[1K\r",
      "Processed 146 (0%) pairs     \u001b[1K\r",
      "Processed 147 (0%) pairs     \u001b[1K\r",
      "Processed 148 (0%) pairs     \u001b[1K\r",
      "Processed 149 (0%) pairs     \u001b[1K\r",
      "Processed 150 (0%) pairs     \u001b[1K\r",
      "Processed 151 (0%) pairs     \u001b[1K\r",
      "Processed 152 (0%) pairs     \u001b[1K\r",
      "Processed 153 (0%) pairs     \u001b[1K\r",
      "Processed 154 (0%) pairs     \u001b[1K\r",
      "Processed 155 (0%) pairs     \u001b[1K\r",
      "Processed 156 (0%) pairs     \u001b[1K\r",
      "Processed 157 (0%) pairs     \u001b[1K\r",
      "Processed 158 (0%) pairs     \u001b[1K\r",
      "Processed 159 (0%) pairs     \u001b[1K\r",
      "Processed 160 (0%) pairs     \u001b[1K\r",
      "Processed 161 (0%) pairs     \u001b[1K\r",
      "Processed 162 (0%) pairs     \u001b[1K\r",
      "Processed 163 (0%) pairs     \u001b[1K\r",
      "Processed 164 (0%) pairs     \u001b[1K\r",
      "Processed 165 (0%) pairs     \u001b[1K\r",
      "Processed 166 (0%) pairs     \u001b[1K\r",
      "Processed 167 (0%) pairs     \u001b[1K\r",
      "Processed 168 (0%) pairs     \u001b[1K\r",
      "Processed 169 (0%) pairs     \u001b[1K\r",
      "Processed 170 (0%) pairs     \u001b[1K\r",
      "Processed 171 (0%) pairs     \u001b[1K\r",
      "Processed 172 (0%) pairs     \u001b[1K\r",
      "Processed 173 (0%) pairs     \u001b[1K\r",
      "Processed 174 (0%) pairs     \u001b[1K\r",
      "Processed 175 (0%) pairs     \u001b[1K\r",
      "Processed 176 (0%) pairs     \u001b[1K\r",
      "Processed 177 (0%) pairs     \u001b[1K\r",
      "Processed 178 (0%) pairs     \u001b[1K\r",
      "Processed 179 (0%) pairs     \u001b[1K\r",
      "Processed 180 (0%) pairs     \u001b[1K\r",
      "Processed 181 (0%) pairs     \u001b[1K\r",
      "Processed 182 (0%) pairs     \u001b[1K\r",
      "Processed 183 (0%) pairs     \u001b[1K\r",
      "Processed 184 (0%) pairs     \u001b[1K\r",
      "Processed 185 (0%) pairs     \u001b[1K\r",
      "Processed 186 (0%) pairs     \u001b[1K\r",
      "Processed 187 (0%) pairs     \u001b[1K\r",
      "Processed 188 (0%) pairs     \u001b[1K\r",
      "Processed 189 (0%) pairs     \u001b[1K\r",
      "Processed 190 (0%) pairs     \u001b[1K\r",
      "Processed 191 (0%) pairs     \u001b[1K\r",
      "Processed 192 (0%) pairs     \u001b[1K\r",
      "Processed 193 (0%) pairs     \u001b[1K\r",
      "Processed 194 (0%) pairs     \u001b[1K\r",
      "Processed 195 (0%) pairs     \u001b[1K\r",
      "Processed 196 (0%) pairs     \u001b[1K\r",
      "Processed 197 (0%) pairs     \u001b[1K\r",
      "Processed 198 (0%) pairs     \u001b[1K\r",
      "Processed 199 (0%) pairs     \u001b[1K\r",
      "Processed 200 (0%) pairs     \u001b[1K\r",
      "Processed 201 (0%) pairs     \u001b[1K\r",
      "Processed 202 (0%) pairs     \u001b[1K\r",
      "Processed 203 (0%) pairs     \u001b[1K\r",
      "Processed 204 (0%) pairs     \u001b[1K\r",
      "Processed 205 (0%) pairs     \u001b[1K\r",
      "Processed 206 (0%) pairs     \u001b[1K\r",
      "Processed 207 (0%) pairs     \u001b[1K\r",
      "Processed 208 (0%) pairs     \u001b[1K\r",
      "Processed 209 (0%) pairs     \u001b[1K\r",
      "Processed 210 (0%) pairs     \u001b[1K\r",
      "Processed 211 (0%) pairs     \u001b[1K\r",
      "Processed 212 (0%) pairs     \u001b[1K\r",
      "Processed 213 (0%) pairs     \u001b[1K\r",
      "Processed 214 (0%) pairs     \u001b[1K\r",
      "Processed 215 (0%) pairs     \u001b[1K\r",
      "Processed 216 (0%) pairs     \u001b[1K\r",
      "Processed 217 (0%) pairs     \u001b[1K\r",
      "Processed 218 (0%) pairs     \u001b[1K\r",
      "Processed 219 (0%) pairs     \u001b[1K\r",
      "Processed 220 (0%) pairs     \u001b[1K\r",
      "Processed 221 (0%) pairs     \u001b[1K\r",
      "Processed 222 (0%) pairs     \u001b[1K\r",
      "Processed 223 (0%) pairs     \u001b[1K\r",
      "Processed 224 (0%) pairs     \u001b[1K\r",
      "Processed 225 (0%) pairs     \u001b[1K\r",
      "Processed 226 (0%) pairs     \u001b[1K\r",
      "Processed 227 (0%) pairs     \u001b[1K\r",
      "Processed 228 (0%) pairs     \u001b[1K\r",
      "Processed 229 (0%) pairs     \u001b[1K\r",
      "Processed 230 (0%) pairs     \u001b[1K\r",
      "Processed 231 (0%) pairs     \u001b[1K\r",
      "Processed 232 (0%) pairs     \u001b[1K\r",
      "Processed 233 (0%) pairs     \u001b[1K\r",
      "Processed 234 (0%) pairs     \u001b[1K\r",
      "Processed 235 (0%) pairs     \u001b[1K\r",
      "Processed 236 (0%) pairs     \u001b[1K\r",
      "Processed 237 (0%) pairs     \u001b[1K\r",
      "Processed 238 (0%) pairs     \u001b[1K\r",
      "Processed 239 (0%) pairs     \u001b[1K\r",
      "Processed 240 (0%) pairs     \u001b[1K\r",
      "Processed 241 (0%) pairs     \u001b[1K\r",
      "Processed 242 (0%) pairs     \u001b[1K\r",
      "Processed 243 (0%) pairs     \u001b[1K\r",
      "Processed 244 (0%) pairs     \u001b[1K\r",
      "Processed 245 (0%) pairs     \u001b[1K\r",
      "Processed 246 (0%) pairs     \u001b[1K\r",
      "Processed 247 (0%) pairs     \u001b[1K\r",
      "Processed 248 (0%) pairs     \u001b[1K\r",
      "Processed 249 (0%) pairs     \u001b[1K\r",
      "Processed 250 (0%) pairs     \u001b[1K\r",
      "Processed 251 (0%) pairs     \u001b[1K\r",
      "Processed 252 (0%) pairs     \u001b[1K\r",
      "Processed 253 (0%) pairs     \u001b[1K\r",
      "Processed 254 (0%) pairs     \u001b[1K\r",
      "Processed 255 (0%) pairs     \u001b[1K\r",
      "Processed 256 (0%) pairs     \u001b[1K\r",
      "Processed 257 (0%) pairs     \u001b[1K\r",
      "Processed 258 (0%) pairs     \u001b[1K\r",
      "Processed 259 (0%) pairs     \u001b[1K\r",
      "Processed 260 (0%) pairs     \u001b[1K\r",
      "Processed 261 (0%) pairs     \u001b[1K\r",
      "Processed 262 (0%) pairs     \u001b[1K\r",
      "Processed 263 (0%) pairs     \u001b[1K\r",
      "Processed 264 (0%) pairs     \u001b[1K\r",
      "Processed 265 (0%) pairs     \u001b[1K\r",
      "Processed 266 (0%) pairs     \u001b[1K\r",
      "Processed 267 (0%) pairs     \u001b[1K\r",
      "Processed 268 (0%) pairs     \u001b[1K\r",
      "Processed 269 (0%) pairs     \u001b[1K\r",
      "Processed 270 (0%) pairs     \u001b[1K\r",
      "Processed 271 (0%) pairs     \u001b[1K\r",
      "Processed 272 (0%) pairs     \u001b[1K\r",
      "Processed 273 (0%) pairs     \u001b[1K\r",
      "Processed 274 (0%) pairs     \u001b[1K\r",
      "Processed 275 (0%) pairs     \u001b[1K\r",
      "Processed 276 (0%) pairs     \u001b[1K\r",
      "Processed 277 (0%) pairs     \u001b[1K\r",
      "Processed 278 (0%) pairs     \u001b[1K\r",
      "Processed 279 (0%) pairs     \u001b[1K\r",
      "Processed 280 (0%) pairs     \u001b[1K\r",
      "Processed 281 (0%) pairs     \u001b[1K\r",
      "Processed 282 (0%) pairs     \u001b[1K\r",
      "Processed 283 (0%) pairs     \u001b[1K\r",
      "Processed 284 (0%) pairs     \u001b[1K\r",
      "Processed 285 (0%) pairs     \u001b[1K\r",
      "Processed 286 (0%) pairs     \u001b[1K\r",
      "Processed 287 (0%) pairs     \u001b[1K\r",
      "Processed 288 (0%) pairs     \u001b[1K\r",
      "Processed 289 (0%) pairs     \u001b[1K\r",
      "Processed 290 (0%) pairs     \u001b[1K\r",
      "Processed 291 (0%) pairs     \u001b[1K\r",
      "Processed 292 (0%) pairs     \u001b[1K\r",
      "Processed 293 (0%) pairs     \u001b[1K\r",
      "Processed 294 (0%) pairs     \u001b[1K\r",
      "Processed 295 (0%) pairs     \u001b[1K\r",
      "Processed 296 (0%) pairs     \u001b[1K\r",
      "Processed 297 (0%) pairs     \u001b[1K\r",
      "Processed 298 (0%) pairs     \u001b[1K\r",
      "Processed 299 (0%) pairs     \u001b[1K\r",
      "Processed 300 (0%) pairs     \u001b[1K\r",
      "Processed 301 (0%) pairs     \u001b[1K\r",
      "Processed 302 (0%) pairs     \u001b[1K\r",
      "Processed 303 (0%) pairs     \u001b[1K\r",
      "Processed 304 (0%) pairs     \u001b[1K\r",
      "Processed 305 (0%) pairs     \u001b[1K\r",
      "Processed 306 (0%) pairs     \u001b[1K\r",
      "Processed 307 (0%) pairs     \u001b[1K\r",
      "Processed 308 (0%) pairs     \u001b[1K\r",
      "Processed 309 (0%) pairs     \u001b[1K\r",
      "Processed 310 (0%) pairs     \u001b[1K\r",
      "Processed 311 (0%) pairs     \u001b[1K\r",
      "Processed 312 (0%) pairs     \u001b[1K\r",
      "Processed 313 (0%) pairs     \u001b[1K\r",
      "Processed 314 (0%) pairs     \u001b[1K\r",
      "Processed 315 (0%) pairs     \u001b[1K\r",
      "Processed 316 (0%) pairs     \u001b[1K\r",
      "Processed 317 (0%) pairs     \u001b[1K\r",
      "Processed 318 (0%) pairs     \u001b[1K\r",
      "Processed 319 (0%) pairs     \u001b[1K\r",
      "Processed 320 (0%) pairs     \u001b[1K\r",
      "Processed 321 (0%) pairs     \u001b[1K\r",
      "Processed 322 (0%) pairs     \u001b[1K\r",
      "Processed 323 (0%) pairs     \u001b[1K\r",
      "Processed 324 (0%) pairs     \u001b[1K\r",
      "Processed 325 (0%) pairs     \u001b[1K\r",
      "Processed 326 (0%) pairs     \u001b[1K\r",
      "Processed 327 (0%) pairs     \u001b[1K\r",
      "Processed 328 (0%) pairs     \u001b[1K\r",
      "Processed 329 (0%) pairs     \u001b[1K\r",
      "Processed 330 (0%) pairs     \u001b[1K\r",
      "Processed 331 (0%) pairs     \u001b[1K\r",
      "Processed 332 (0%) pairs     \u001b[1K\r",
      "Processed 333 (0%) pairs     \u001b[1K\r",
      "Processed 334 (0%) pairs     \u001b[1K\r",
      "Processed 335 (0%) pairs     \u001b[1K\r",
      "Processed 336 (0%) pairs     \u001b[1K\r",
      "Processed 337 (0%) pairs     \u001b[1K\r",
      "Processed 338 (0%) pairs     \u001b[1K\r",
      "Processed 339 (0%) pairs     \u001b[1K\r",
      "Processed 340 (0%) pairs     \u001b[1K\r",
      "Processed 341 (0%) pairs     \u001b[1K\r",
      "Processed 342 (0%) pairs     \u001b[1K\r",
      "Processed 343 (0%) pairs     \u001b[1K\r",
      "Processed 344 (0%) pairs     \u001b[1K\r",
      "Processed 345 (0%) pairs     \u001b[1K\r",
      "Processed 346 (0%) pairs     \u001b[1K\r",
      "Processed 347 (0%) pairs     \u001b[1K\r",
      "Processed 348 (0%) pairs     \u001b[1K\r",
      "Processed 349 (0%) pairs     \u001b[1K\r",
      "Processed 350 (0%) pairs     \u001b[1K\r",
      "Processed 351 (0%) pairs     \u001b[1K\r",
      "Processed 352 (0%) pairs     \u001b[1K\r",
      "Processed 353 (0%) pairs     \u001b[1K\r",
      "Processed 354 (0%) pairs     \u001b[1K\r",
      "Processed 355 (0%) pairs     \u001b[1K\r",
      "Processed 356 (0%) pairs     \u001b[1K\r",
      "Processed 357 (0%) pairs     \u001b[1K\r",
      "Processed 358 (0%) pairs     \u001b[1K\r",
      "Processed 359 (0%) pairs     \u001b[1K\r",
      "Processed 360 (0%) pairs     \u001b[1K\r",
      "Processed 361 (0%) pairs     \u001b[1K\r",
      "Processed 362 (0%) pairs     \u001b[1K\r",
      "Processed 363 (0%) pairs     \u001b[1K\r",
      "Processed 364 (0%) pairs     \u001b[1K\r",
      "Processed 365 (0%) pairs     \u001b[1K\r",
      "Processed 366 (0%) pairs     \u001b[1K\r",
      "Processed 367 (0%) pairs     \u001b[1K\r",
      "Processed 368 (0%) pairs     \u001b[1K\r",
      "Processed 369 (0%) pairs     \u001b[1K\r",
      "Processed 370 (0%) pairs     \u001b[1K\r",
      "Processed 371 (0%) pairs     \u001b[1K\r",
      "Processed 372 (0%) pairs     \u001b[1K\r",
      "Processed 373 (0%) pairs     \u001b[1K\r",
      "Processed 374 (0%) pairs     \u001b[1K\r",
      "Processed 375 (0%) pairs     \u001b[1K\r",
      "Processed 376 (0%) pairs     \u001b[1K\r",
      "Processed 377 (0%) pairs     \u001b[1K\r",
      "Processed 378 (0%) pairs     \u001b[1K\r",
      "Processed 379 (0%) pairs     \u001b[1K\r",
      "Processed 380 (0%) pairs     \u001b[1K\r",
      "Processed 381 (0%) pairs     \u001b[1K\r",
      "Processed 382 (0%) pairs     \u001b[1K\r",
      "Processed 383 (0%) pairs     \u001b[1K\r",
      "Processed 384 (0%) pairs     \u001b[1K\r",
      "Processed 385 (0%) pairs     \u001b[1K\r",
      "Processed 386 (0%) pairs     \u001b[1K\r",
      "Processed 387 (0%) pairs     \u001b[1K\r",
      "Processed 388 (0%) pairs     \u001b[1K\r",
      "Processed 389 (0%) pairs     \u001b[1K\r",
      "Processed 390 (0%) pairs     \u001b[1K\r",
      "Processed 391 (0%) pairs     \u001b[1K\r",
      "Processed 392 (0%) pairs     \u001b[1K\r",
      "Processed 393 (0%) pairs     \u001b[1K\r",
      "Processed 394 (0%) pairs     \u001b[1K\r",
      "Processed 395 (0%) pairs     \u001b[1K\r",
      "Processed 396 (0%) pairs     \u001b[1K\r",
      "Processed 397 (0%) pairs     \u001b[1K\r",
      "Processed 398 (0%) pairs     \u001b[1K\r",
      "Processed 399 (0%) pairs     \u001b[1K\r",
      "Processed 400 (0%) pairs     \u001b[1K\r",
      "Processed 401 (0%) pairs     \u001b[1K\r",
      "Processed 402 (0%) pairs     \u001b[1K\r",
      "Processed 403 (0%) pairs     \u001b[1K\r",
      "Processed 404 (0%) pairs     \u001b[1K\r",
      "Processed 405 (0%) pairs     \u001b[1K\r",
      "Processed 406 (0%) pairs     \u001b[1K\r",
      "Processed 407 (0%) pairs     \u001b[1K\r",
      "Processed 408 (0%) pairs     \u001b[1K\r",
      "Processed 409 (0%) pairs     \u001b[1K\r",
      "Processed 410 (0%) pairs     \u001b[1K\r",
      "Processed 411 (0%) pairs     \u001b[1K\r",
      "Processed 412 (0%) pairs     \u001b[1K\r",
      "Processed 413 (0%) pairs     \u001b[1K\r",
      "Processed 414 (0%) pairs     \u001b[1K\r",
      "Processed 415 (0%) pairs     \u001b[1K\r",
      "Processed 416 (0%) pairs     \u001b[1K\r",
      "Processed 417 (0%) pairs     \u001b[1K\r",
      "Processed 418 (0%) pairs     \u001b[1K\r",
      "Processed 419 (0%) pairs     \u001b[1K\r",
      "Processed 420 (0%) pairs     \u001b[1K\r",
      "Processed 421 (0%) pairs     \u001b[1K\r",
      "Processed 422 (0%) pairs     \u001b[1K\r",
      "Processed 423 (0%) pairs     \u001b[1K\r",
      "Processed 424 (0%) pairs     \u001b[1K\r",
      "Processed 425 (0%) pairs     \u001b[1K\r",
      "Processed 426 (0%) pairs     \u001b[1K\r",
      "Processed 427 (0%) pairs     \u001b[1K\r",
      "Processed 428 (0%) pairs     \u001b[1K\r",
      "Processed 429 (0%) pairs     \u001b[1K\r",
      "Processed 430 (0%) pairs     \u001b[1K\r",
      "Processed 431 (0%) pairs     \u001b[1K\r",
      "Processed 432 (0%) pairs     \u001b[1K\r",
      "Processed 433 (0%) pairs     \u001b[1K\r",
      "Processed 434 (0%) pairs     \u001b[1K\r",
      "Processed 435 (0%) pairs     \u001b[1K\r",
      "Processed 436 (0%) pairs     \u001b[1K\r",
      "Processed 437 (0%) pairs     \u001b[1K\r",
      "Processed 438 (0%) pairs     \u001b[1K\r",
      "Processed 439 (0%) pairs     \u001b[1K\r",
      "Processed 440 (0%) pairs     \u001b[1K\r",
      "Processed 441 (0%) pairs     \u001b[1K\r",
      "Processed 442 (0%) pairs     \u001b[1K\r",
      "Processed 443 (0%) pairs     \u001b[1K\r",
      "Processed 444 (0%) pairs     \u001b[1K\r",
      "Processed 445 (0%) pairs     \u001b[1K\r",
      "Processed 446 (0%) pairs     \u001b[1K\r",
      "Processed 447 (0%) pairs     \u001b[1K\r",
      "Processed 448 (0%) pairs     \u001b[1K\r",
      "Processed 449 (0%) pairs     \u001b[1K\r",
      "Processed 450 (0%) pairs     \u001b[1K\r",
      "Processed 451 (0%) pairs     \u001b[1K\r",
      "Processed 452 (0%) pairs     \u001b[1K\r",
      "Processed 453 (0%) pairs     \u001b[1K\r",
      "Processed 454 (0%) pairs     \u001b[1K\r",
      "Processed 455 (0%) pairs     \u001b[1K\r",
      "Processed 456 (0%) pairs     \u001b[1K\r",
      "Processed 457 (0%) pairs     \u001b[1K\r",
      "Processed 458 (0%) pairs     \u001b[1K\r",
      "Processed 459 (0%) pairs     \u001b[1K\r",
      "Processed 460 (0%) pairs     \u001b[1K\r",
      "Processed 461 (0%) pairs     \u001b[1K\r",
      "Processed 462 (0%) pairs     \u001b[1K\r",
      "Processed 463 (0%) pairs     \u001b[1K\r",
      "Processed 464 (0%) pairs     \u001b[1K\r",
      "Processed 465 (0%) pairs     \u001b[1K\r",
      "Processed 466 (0%) pairs     \u001b[1K\r",
      "Processed 467 (0%) pairs     \u001b[1K\r",
      "Processed 468 (0%) pairs     \u001b[1K\r",
      "Processed 469 (0%) pairs     \u001b[1K\r",
      "Processed 470 (0%) pairs     \u001b[1K\r",
      "Processed 471 (0%) pairs     \u001b[1K\r",
      "Processed 472 (0%) pairs     \u001b[1K\r",
      "Processed 473 (0%) pairs     \u001b[1K\r",
      "Processed 474 (0%) pairs     \u001b[1K\r",
      "Processed 475 (0%) pairs     \u001b[1K\r",
      "Processed 476 (0%) pairs     \u001b[1K\r",
      "Processed 477 (0%) pairs     \u001b[1K\r",
      "Processed 478 (0%) pairs     \u001b[1K\r",
      "Processed 479 (0%) pairs     \u001b[1K\r",
      "Processed 480 (0%) pairs     \u001b[1K\r",
      "Processed 481 (0%) pairs     \u001b[1K\r",
      "Processed 482 (0%) pairs     \u001b[1K\r",
      "Processed 483 (0%) pairs     \u001b[1K\r",
      "Processed 484 (0%) pairs     \u001b[1K\r",
      "Processed 485 (0%) pairs     \u001b[1K\r",
      "Processed 486 (0%) pairs     \u001b[1K\r",
      "Processed 487 (0%) pairs     \u001b[1K\r",
      "Processed 488 (0%) pairs     \u001b[1K\r",
      "Processed 489 (0%) pairs     \u001b[1K\r",
      "Processed 490 (0%) pairs     \u001b[1K\r",
      "Processed 491 (0%) pairs     \u001b[1K\r",
      "Processed 492 (0%) pairs     \u001b[1K\r",
      "Processed 493 (0%) pairs     \u001b[1K\r",
      "Processed 494 (0%) pairs     \u001b[1K\r",
      "Processed 495 (0%) pairs     \u001b[1K\r",
      "Processed 496 (0%) pairs     \u001b[1K\r",
      "Processed 497 (0%) pairs     \u001b[1K\r",
      "Processed 498 (0%) pairs     \u001b[1K\r",
      "Processed 499 (0%) pairs     \u001b[1K\r",
      "Processed 500 (0%) pairs     \u001b[1K\r",
      "Processed 501 (0%) pairs     \u001b[1K\r",
      "Processed 502 (0%) pairs     \u001b[1K\r",
      "Processed 503 (0%) pairs     \u001b[1K\r",
      "Processed 504 (0%) pairs     \u001b[1K\r",
      "Processed 505 (0%) pairs     \u001b[1K\r",
      "Processed 506 (0%) pairs     \u001b[1K\r",
      "Processed 507 (0%) pairs     \u001b[1K\r",
      "Processed 508 (0%) pairs     \u001b[1K\r",
      "Processed 509 (0%) pairs     \u001b[1K\r",
      "Processed 510 (0%) pairs     \u001b[1K\r",
      "Processed 511 (0%) pairs     \u001b[1K\r",
      "Processed 512 (0%) pairs     \u001b[1K\r",
      "Processed 513 (0%) pairs     \u001b[1K\r",
      "Processed 514 (0%) pairs     \u001b[1K\r",
      "Processed 515 (0%) pairs     \u001b[1K\r",
      "Processed 516 (0%) pairs     \u001b[1K\r",
      "Processed 517 (0%) pairs     \u001b[1K\r",
      "Processed 518 (0%) pairs     \u001b[1K\r",
      "Processed 519 (0%) pairs     \u001b[1K\r",
      "Processed 520 (0%) pairs     \u001b[1K\r",
      "Processed 521 (0%) pairs     \u001b[1K\r",
      "Processed 522 (0%) pairs     \u001b[1K\r",
      "Processed 523 (0%) pairs     \u001b[1K\r",
      "Processed 524 (0%) pairs     \u001b[1K\r",
      "Processed 525 (0%) pairs     \u001b[1K\r",
      "Processed 526 (0%) pairs     \u001b[1K\r",
      "Processed 527 (0%) pairs     \u001b[1K\r",
      "Processed 528 (0%) pairs     \u001b[1K\r",
      "Processed 529 (0%) pairs     \u001b[1K\r",
      "Processed 530 (0%) pairs     \u001b[1K\r",
      "Processed 531 (0%) pairs     \u001b[1K\r",
      "Processed 532 (0%) pairs     \u001b[1K\r",
      "Processed 533 (0%) pairs     \u001b[1K\r",
      "Processed 534 (0%) pairs     \u001b[1K\r",
      "Processed 535 (0%) pairs     \u001b[1K\r",
      "Processed 536 (0%) pairs     \u001b[1K\r",
      "Processed 537 (0%) pairs     \u001b[1K\r",
      "Processed 538 (0%) pairs     \u001b[1K\r",
      "Processed 539 (0%) pairs     \u001b[1K\r",
      "Processed 540 (0%) pairs     \u001b[1K\r",
      "Processed 541 (0%) pairs     \u001b[1K\r",
      "Processed 542 (0%) pairs     \u001b[1K\r",
      "Processed 543 (0%) pairs     \u001b[1K\r",
      "Processed 544 (0%) pairs     \u001b[1K\r",
      "Processed 545 (0%) pairs     \u001b[1K\r",
      "Processed 546 (0%) pairs     \u001b[1K\r",
      "Processed 547 (0%) pairs     \u001b[1K\r",
      "Processed 548 (0%) pairs     \u001b[1K\r",
      "Processed 549 (0%) pairs     \u001b[1K\r",
      "Processed 550 (0%) pairs     \u001b[1K\r",
      "Processed 551 (0%) pairs     \u001b[1K\r",
      "Processed 552 (0%) pairs     \u001b[1K\r",
      "Processed 553 (0%) pairs     \u001b[1K\r",
      "Processed 554 (0%) pairs     \u001b[1K\r",
      "Processed 555 (0%) pairs     \u001b[1K\r",
      "Processed 556 (0%) pairs     \u001b[1K\r",
      "Processed 557 (0%) pairs     \u001b[1K\r",
      "Processed 558 (0%) pairs     \u001b[1K\r",
      "Processed 559 (0%) pairs     \u001b[1K\r",
      "Processed 560 (0%) pairs     \u001b[1K\r",
      "Processed 561 (0%) pairs     \u001b[1K\r",
      "Processed 562 (0%) pairs     \u001b[1K\r",
      "Processed 563 (0%) pairs     \u001b[1K\r",
      "Processed 564 (0%) pairs     \u001b[1K\r",
      "Processed 565 (0%) pairs     \u001b[1K\r",
      "Processed 566 (0%) pairs     \u001b[1K\r",
      "Processed 567 (0%) pairs     \u001b[1K\r",
      "Processed 568 (0%) pairs     \u001b[1K\r",
      "Processed 569 (0%) pairs     \u001b[1K\r",
      "Processed 570 (0%) pairs     \u001b[1K\r",
      "Processed 571 (0%) pairs     \u001b[1K\r",
      "Processed 572 (0%) pairs     \u001b[1K\r",
      "Processed 573 (0%) pairs     \u001b[1K\r",
      "Processed 574 (0%) pairs     \u001b[1K\r",
      "Processed 575 (0%) pairs     \u001b[1K\r",
      "Processed 576 (0%) pairs     \u001b[1K\r",
      "Processed 577 (0%) pairs     \u001b[1K\r",
      "Processed 578 (0%) pairs     \u001b[1K\r",
      "Processed 579 (0%) pairs     \u001b[1K\r",
      "Processed 580 (0%) pairs     \u001b[1K\r",
      "Processed 581 (0%) pairs     \u001b[1K\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reached the edge limit (1000)  \u001b[1K\n",
      "Generated 1000 edges                    \n",
      "Created a co-occurrence graph:\n",
      "\tnumber of nodes:  896\n",
      "\tnumber of edges:  1000\n",
      "Saving the edges...\n",
      "Creating a graph object...\n"
     ]
    }
   ],
   "source": [
    "# Limit to 1000 edges\n",
    "paper_comention_network_1000_edges = generate_comention_network(\n",
    "    occurrence_data, counts[\"paper\"],\n",
    "    factor_column=\"paper\",\n",
    "#     n_most_frequent=100,\n",
    "    limit=1000,\n",
    "    parallelize=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitering data.....\n",
      "Selected 100 most frequent terms\n",
      "Examining 4950 pairs of terms for co-occurrence...\n",
      "Generated 3421 edges                    \n",
      "Created a co-occurrence graph:\n",
      "\tnumber of nodes:  100\n",
      "\tnumber of edges:  3421\n",
      "Saving the edges...\n",
      "Creating a graph object...\n"
     ]
    }
   ],
   "source": [
    "# Use only 100 most frequent entities\n",
    "paper_comention_network_100_most_frequent = generate_comention_network(\n",
    "    occurrence_data, counts[\"paper\"],\n",
    "    factor_column=\"paper\",\n",
    "    n_most_frequent=100,\n",
    "#     limit=1000,\n",
    "    parallelize=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Compute centralities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We compute the degree and PageRank centralities only for the raw frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "weights = [\"frequency\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top n nodes by frequency:\n",
      "\thuman (420)\n",
      "\tvirus (353)\n",
      "\thumans (312)\n",
      "\tviral (284)\n",
      "\tlung (283)\n",
      "\tinfection (282)\n",
      "\tdna (266)\n",
      "\tards (256)\n",
      "\tbacterial (256)\n",
      "\tanimal (255)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "degree_centrality = compute_degree_centrality(paper_comention_network_100_most_frequent, weights, 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top n nodes by frequency:\n",
      "\thuman (0.03)\n",
      "\tvirus (0.02)\n",
      "\thumans (0.02)\n",
      "\tinfection (0.02)\n",
      "\tviral (0.02)\n",
      "\tlung (0.02)\n",
      "\tdna (0.02)\n",
      "\tbacterial (0.02)\n",
      "\tanimal (0.02)\n",
      "\tanimals (0.02)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pagerank_centrality = compute_pagerank_centrality(paper_comention_network_100_most_frequent, weights, 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We then compute the betweenness centrality based on the PPMI and NPMI distances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top n nodes by distance_ppmi:\n",
      "\tcell lines (0.06547447261732976)\n",
      "\talveolar macrophages (0.047124549675570074)\n",
      "\tamino acid sequence (0.04497354497354497)\n",
      "\tco2 (0.04057926200783344)\n",
      "\tantioxidants (0.033897036448056854)\n",
      "\taids (0.021713735999450284)\n",
      "\tepithelial cells (0.02170023853697323)\n",
      "\thuman disease (0.021438878581735724)\n",
      "\tasthmatic (0.021208930902808454)\n",
      "\tallergic (0.01941180512609084)\n",
      "\theart (0.018793375936233075)\n",
      "\tbacterium (0.01742596028310314)\n",
      "\tamino acids (0.015838658695801554)\n",
      "\ttranscription factors (0.01504844361987219)\n",
      "\tamino acid (0.014979729265443548)\n",
      "\tcystic fibrosis (0.01392324606610321)\n",
      "\trats (0.013399299113584827)\n",
      "\tformaldehyde (0.013090084518655946)\n",
      "\tenzyme (0.012677798392084105)\n",
      "\tinfected cells (0.012358276643990931)\n",
      "\n",
      "Top n nodes by distance_npmi:\n",
      "\ttranscription factors (0.03150553150553151)\n",
      "\tmouse (0.0282415996701711)\n",
      "\tamino acids (0.027004741290455575)\n",
      "\theart (0.025613275613275612)\n",
      "\tamino acid (0.024324881467738608)\n",
      "\thuman disease (0.023809523809523808)\n",
      "\tepithelial cells (0.023088023088023088)\n",
      "\tcell lines (0.02198859341716484)\n",
      "\ttumor (0.020476877619734764)\n",
      "\tinfected cells (0.02027760599189171)\n",
      "\tcystic fibrosis (0.02023637737923452)\n",
      "\tco2 (0.018844911702054556)\n",
      "\talveolar macrophages (0.017934446505875078)\n",
      "\tmolecules (0.016903731189445475)\n",
      "\tbovine (0.016663230948945233)\n",
      "\trat (0.016388373531230672)\n",
      "\tenzyme (0.015426372569229711)\n",
      "\tneutrophils (0.01468769325912183)\n",
      "\tairways (0.01443001443001443)\n",
      "\tmurine (0.013914656771799629)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "betweenness_centrality = compute_betweenness_centrality(\n",
    "    paper_comention_network_100_most_frequent, [\"distance_ppmi\", \"distance_npmi\"], 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Detect communities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detecting communities...\n",
      "Best network partition:\n",
      "\t Number of communities: 2\n",
      "\t Modularity: 0.16723090179549283\n",
      "Detecting communities...\n",
      "Best network partition:\n",
      "\t Number of communities: 3\n",
      "\t Modularity: 0.22420884837314267\n",
      "Detecting communities...\n",
      "Best network partition:\n",
      "\t Number of communities: 2\n",
      "\t Modularity: 0.24539789862976222\n"
     ]
    }
   ],
   "source": [
    "_ = detect_communities(paper_comention_network_100_most_frequent, weight=\"frequency\", set_attr=\"community\")\n",
    "_ = detect_communities(paper_comention_network_100_most_frequent, weight=\"ppmi\", set_attr=\"community_ppmi\")\n",
    "_ = detect_communities(paper_comention_network_100_most_frequent, weight=\"npmi\", set_attr=\"community_npmi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Compute all the metrics in one go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitering data.....\n",
      "Selected 100 most frequent terms\n",
      "Examining 4950 pairs of terms for co-occurrence...\n",
      "Generated 3421 edges                    \n",
      "Created a co-occurrence graph:\n",
      "\tnumber of nodes:  100\n",
      "\tnumber of edges:  3421\n",
      "Saving the edges...\n",
      "Creating a graph object...\n"
     ]
    }
   ],
   "source": [
    "paper_comention_network = generate_comention_network(\n",
    "    occurrence_data, counts[\"paper\"],\n",
    "    factor_column=\"paper\",\n",
    "    n_most_frequent=100,\n",
    "#     limit=1000,\n",
    "    parallelize=False,\n",
    "    dump_path=\"data/paper_comention_edge_list.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing degree centrality statistics....\n",
      "Top n nodes by frequency:\n",
      "\thuman (420)\n",
      "\tvirus (353)\n",
      "\thumans (312)\n",
      "\tviral (284)\n",
      "\tlung (283)\n",
      "\tinfection (282)\n",
      "\tdna (266)\n",
      "\tards (256)\n",
      "\tbacterial (256)\n",
      "\tanimal (255)\n",
      "\n",
      "Computing PageRank centrality statistics....\n",
      "Top n nodes by frequency:\n",
      "\thuman (0.03)\n",
      "\tvirus (0.02)\n",
      "\thumans (0.02)\n",
      "\tinfection (0.02)\n",
      "\tviral (0.02)\n",
      "\tlung (0.02)\n",
      "\tdna (0.02)\n",
      "\tbacterial (0.02)\n",
      "\tanimal (0.02)\n",
      "\tanimals (0.02)\n",
      "\n",
      "Computing betweenness centrality statistics....\n",
      "Top n nodes by distance_ppmi:\n",
      "\tcell lines (0.06547447261732976)\n",
      "\talveolar macrophages (0.047124549675570074)\n",
      "\tamino acid sequence (0.04497354497354497)\n",
      "\tco2 (0.04057926200783344)\n",
      "\tantioxidants (0.033897036448056854)\n",
      "\taids (0.021713735999450284)\n",
      "\tepithelial cells (0.02170023853697323)\n",
      "\thuman disease (0.021438878581735724)\n",
      "\tasthmatic (0.021208930902808454)\n",
      "\tallergic (0.01941180512609084)\n",
      "\n",
      "Top n nodes by distance_npmi:\n",
      "\ttranscription factors (0.03150553150553151)\n",
      "\tmouse (0.0282415996701711)\n",
      "\tamino acids (0.027004741290455575)\n",
      "\theart (0.025613275613275612)\n",
      "\tamino acid (0.024324881467738608)\n",
      "\thuman disease (0.023809523809523808)\n",
      "\tepithelial cells (0.023088023088023088)\n",
      "\tcell lines (0.02198859341716484)\n",
      "\ttumor (0.020476877619734764)\n",
      "\tinfected cells (0.02027760599189171)\n",
      "\n",
      "Detecting communities...\n",
      "Best network partition:\n",
      "\t Number of communities: 2\n",
      "\t Modularity: 0.17062247895364224\n",
      "Detecting communities...\n",
      "Best network partition:\n",
      "\t Number of communities: 2\n",
      "\t Modularity: 0.24539789862976222\n"
     ]
    }
   ],
   "source": [
    "# We need to specify edge attributes used for computing different metrics\n",
    "compute_all_metrics(\n",
    "    paper_comention_network,\n",
    "    degree_weights=[\"frequency\"],\n",
    "    pagerank_weights=[\"frequency\"],\n",
    "    betweenness_weights=[\"distance_ppmi\", \"distance_npmi\"],\n",
    "    community_weights=[\"frequency\", \"npmi\"],\n",
    "    print_summary=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Export network and the computed metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save graph nodes as a pickled pandas.DataFrame\n",
    "save_nodes(paper_comention_network, \"data/paper_comention_node_list.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'degree_frequency': 244,\n",
       " 'pagerank_frequency': 0.015405249811721203,\n",
       " 'betweenness_distance_ppmi': 0.0,\n",
       " 'betweenness_distance_npmi': 0.002886002886002886,\n",
       " 'community_frequency': 0,\n",
       " 'community_npmi': 0}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paper_comention_network.nodes[list(paper_comention_network.nodes())[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the graph for Gephi import.\n",
    "save_to_gephi(\n",
    "    paper_comention_network, \"data/gephi_paper\", \n",
    "    node_attr_mapping = {\n",
    "        \"degree_frequency\": \"Degree\",\n",
    "        \"pagerank_frequency\": \"PageRank\",\n",
    "        \"betweenness_distance_npmi\": \"Betweenness\",\n",
    "        \"community_npmi\": \"Community\"\n",
    "    },\n",
    "    edge_attr_mapping={\n",
    "        \"npmi\": \"Weight\"\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Find spanning trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = minimum_spanning_tree(paper_comention_network, weight=\"distance_npmi\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_to_gephi(\n",
    "    tree, \"data/gephi_paper_spanning_tree\", \n",
    "    node_attr_mapping = {\n",
    "        \"degree_frequency\": \"Degree\",\n",
    "        \"pagerank_frequency\": \"PageRank\",\n",
    "        \"betweenness_distance_npmi\": \"Betweenness\",\n",
    "        \"community_npmi\": \"Community\"\n",
    "    },\n",
    "    edge_attr_mapping={\n",
    "        \"npmi\": \"Weight\"\n",
    "    })"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Path search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(\"Detecting communities....\")\n",
    "# detect_communities(graph, weight=\"frequency\", set_attr=\"community\")\n",
    "# detect_communities(graph, weight=\"ppmi\", set_attr=\"community_ppmi\")\n",
    "# detect_communities(graph, weight=\"npmi\", set_attr=\"community_npmi\")\n",
    "\n",
    "# print(\"Saving the nodes...\")\n",
    "# save_nodes(graph, os.path.join(dump_path, \"{}_{}_node_list.pkl\".format(factor, N)))\n",
    "\n",
    "# print(\"Computing spanning trees...\")\n",
    "# tree_ppmi = nx.minimum_spanning_tree(graph, weight=\"distance_ppmi\")\n",
    "# tree_npmi = nx.minimum_spanning_tree(graph, weight=\"distance_npmi\")\n",
    "\n",
    "# print(\"Saving spanning trees...\")\n",
    "# edges = tree_ppmi.edges()\n",
    "# save_edges(\n",
    "#     tree_ppmi, os.path.join(dump_path, \"{}_{}_ppmi_spanning_tree.pkl\".format(factor, N)))\n",
    "# save_edges(\n",
    "#     tree_npmi, os.path.join(dump_path, \"{}_{}_npmi_spanning_tree.pkl\".format(factor, N)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bbg",
   "language": "python",
   "name": "bbg"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
